blockchain/chain_sync/src/bucket.rs

```diff
@@ -57,9 +60,18 @@ impl SyncBucketSet {
         }
         self.buckets.push(SyncBucket::new(vec![tipset]))
     }
-    pub(crate) fn heaviest(&self) -> Option<Arc<Tipset>> {
+    /// Removes the SyncBucket with heaviest weighted Tipset from SyncBucketSet
+    pub(crate) fn pop(&mut self) -> Option<SyncBucket> {
+        if let Some(heaviest_bucket) = self.buckets().iter().max_by_key(|b| b.heaviest_tipset()) {
+            self.clone().remove(heaviest_bucket);
```

Comment:
```suggestion
            self.remove(heaviest_bucket);
```

Reply:
Need to clone, the compiler complains otherwise.

Reply:
That's because the logic of the function is flawed, you need to be mutating self to remove the actual value or this will just create a copy of the removed item. Also the `Some()` return doesn't need to clone if you keep similar logic

Reply:
you can do this instead: 
```
        if let Some(heaviest_bucket) = self.buckets().iter().max_by_key(|b| b.heaviest_tipset()).clone() {
            let heaviest_bucket = heaviest_bucket.clone();
            self.remove(&heaviest_bucket);
            Some(heaviest_bucket)
        } else {
            None
        }
```

Reply:
That still requires a clone of a bucket of tipsets, better to not rely on this

Reply:
oh youre right... i didnt realize it was a whole bunch of tipsets

Reply:
One option, as I mentioned:
```
        if let Some((i, _)) = self
            .buckets()
            .iter()
            .enumerate()
            .max_by_key(|(_, b)| b.heaviest_tipset())
        {
            let ts = self.buckets.remove(i);
            Some(ts)
        } else {
            None
        }
```
you are already doing all of the checks for finding the max, seems redundant to have the remove function anyway. This would obviously have to be tested, just wanted to put my thought into something tangible. 

I don't think the max_by_key is terribly efficient, especially since the logic that was changed in this PR has heaviest_tipset cloning that heaviest tipset, (sub thought is that you don't clone the tipset and order based on that as a key as that may not even order correctly since it compares the tipset not the weight)